### PR TITLE
Infra/messages rework

### PR DESCRIPTION
**Pull request by woonmoon**
### **Overview**
This should resolve _most of_ [Issue 40](https://github.com/SOMAS2021/SOMAS2021/issues/40). The proposed message types can be divided into 4 categories
- **`AskMessage`**
Messages that inquires the state of neighboring agents
e.g. `askIdentity` (the "Who are you?" message) or `askIntendedFoodIntake` ("How much food are you going to take?")
- **`StateMessage`**
Messages that announce something about the state of an agent
e.g. `stateIdentity` (the "I am..." message) or `stateIntendedFoodIntake` ("This is how much food I'm intending to take...")
- **`RequestMessage`**
Messages that request something from neighboring agents
e.g. `requestLeaveFood` ("please leave x amount of food on the platform")
- **`ResponseMessage`**
A message indicating either agreement or disagreement.

Notes:
- `StateMessage` is the valid response to an `AskMessage` 
- `StateMessage` is also a valid message on its own
- `ResponseMessage` is the valid response to an `RequestMessage` 
- `ResponseMessage` is _not_ valid messages on its own
- Agents should be allowed to customize their handling of every one of these messages, including ignoring entirely (not processing the received message at all) or extracting the information from a message (or note that the message has been sent) and not react/reply. 

### Architectural Changes
The `Agent` interface has been moved from `globalTypes` package to `infra`, this was to avoid circular dependencies between the messages and agents.
`BaseMessage` on its own is no longer a valid message type to transmit across agents (it does not have a defined `MessageType`, for teams that implemented `BaseMessage` transmissions I edited them to defined message types. Team 3 and Team 6 were affected)
`Agent` interface now implements `Handle` functions for each of the message types. 
**All future message types will need their own handler in the Agent interface and all custom agents will need to implement something (even an empty handler) for the new message type.**
Visitor-pattern was used as opposed to a master switch-case statement for readability, scalability, and safety (this ensures that no agent would ever receive a message they can not parse).

### API
For a custom agent to react to a received message it must call `message.Visit(agent)` like below (snippet from `team1/agent1.go`'s `Run()`:
```
receivedMsg := a.Base.ReceiveMessage()
if receivedMsg != nil {
	receivedMsg.Visit(a)
} else {
	a.Log("I got no thing")
}
```
The `Visit` function will trigger the correct handler on the agent depending on the message type.
`AskMessage` and `RequestMessage` implement `Reply(...)` which given some parameters will construct and return the matching reply message (e.g. `AskFoodTaken`'s reply will be `StateFoodTaken`).
`StateMessage` `RequestMessage` and `ResponseMessage` implement `Statement()`, `Request()` and `Response()` respectively which returns the numeric values under the messages (e.g. `Statement()` on a `StateFoodTakenMessage(..., 20)` struct will return 20), or the `Response()` of a disagreeing `ResponseMessage` will return false) (these will probably be useful in the custom agent's handling functions)

### Simulation
The current simulation on this PR simulates `team1/agent1` and `team1/agent2` on floors 1 and 2 respectively. `agent1` sends a randomly chosen message to `agent2` and may also receive replies. 
For examples of basic handler functions, please read the implementations under `team1` and try running/tracing the simulation output with this config.

### Conclusion
 This is not a complete implementation, this is a framework on how the common language should function and how we can enforce safe communication with the existing message proposals. Some of the message types have not been written and unique message IDs have not been implemented yet. However, I wanted to PR this semi-WIP so more people could pitch in on message implementation and for agent teams to begin writing their handlers.
Also, it's 3am as I PR this so this is definitely going to be sloppy... :')

### CODE REVIEW

File: pkg/simulation/simAgent.go [hunk 1]
```diff
@@ -28,7 +24,8 @@ func (sE *SimEnv) generateInitialAgents(t *infra.Tower) {
 func (sE *SimEnv) createNewAgent(tower *infra.Tower, i, floor int) {
 	// TODO: clean this looping, make a nice abs map
 	sE.Log("Creating new agent")
-	abs := []AgentNewFunc{agent1.New, agent2.New, team3.New, team6.New, randomAgent.New}
+	// abs := []AgentNewFunc{agent1.New, agent2.New, team3.New, team6.New, randomAgent.New}
```
[the remark this answers is not in the document]
Comment (hussain2603) [reply pkg/simulation/simAgent.go:27]: just changing it back to what it was before

**Review by hussain2603 — CHANGES_REQUESTED** [review]
overall looks good to me.

File: pkg/messages/askFoodTakenMessage.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+func (msg *AskFoodTakenMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *AskFoodTakenMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
+}
+
+func (msg *AskFoodTakenMessage) Visit(a infra.Agent) {
```
Comment (hussain2603) [inline pkg/messages/askFoodTakenMessage.go:31]: Can we call 'Visit' something else pls? 'CallToHandler' maybe

Comment (jzzheng22) [reply pkg/messages/askFoodTakenMessage.go:31]: https://refactoring.guru/design-patterns/visitor/go/example

It's a Go thing

**Review by jzzheng22 — DISMISSED** [review]
I think this is good, thanks so much

File: cmd/backend/main.go [hunk 3]
```diff
@@ -37,7 +37,7 @@ func main() {
 
 	// can have frontend parameters come go straight into simEnv
 	foodOnPlatform := 100.0
-	numOfAgents := []int{2, 2, 2, 2, 2} //agent1, agent2, team3, team6, randomAgent
+	numOfAgents := []int{1, 1}
```
Comment (jzzheng22) [inline cmd/backend/main.go:40]:
```suggestion
	numOfAgents := []int{2, 2, 2, 2, 2} //agent1, agent2, team3, team6, randomAgent
```

File: pkg/agents/team3/agentMain.go [hunk 4]
```diff
@@ -31,8 +29,8 @@ type CustomAgent3 struct {
 	//and an array of tuples for friendships
 }
 
-func New(baseAgent *infra.Base) (agent.Agent, error) {
-
+func New(baseAgent *infra.Base) (infra.Agent, error) {
+	// TODO: Remove this line. See Issue #60.
```
Comment (jzzheng22) [inline pkg/agents/team3/agentMain.go:33]:
```suggestion
```

File: pkg/simulation/simAgent.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jzzheng22) [reply pkg/simulation/simAgent.go:27]: Agree with the above

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

File: pkg/agents/team1/agent1/agent.go [hunk 5]
```diff
@@ -23,19 +24,100 @@ func (a *CustomAgent1) Run() {
 
 	receivedMsg := a.Base.ReceiveMessage()
```
Comment (JaafarRammal) [inline pkg/agents/team1/agent1/agent.go:25]:
```suggestion
	receivedMsg := a.ReceiveMessage()
```

File: pkg/agents/team1/agent2/agent.go [hunk 6]
```diff
@@ -22,4 +21,61 @@ func New(baseAgent *infra.Base) (agent.Agent, error) {
 func (a *CustomAgent2) Run() {
 	a.Log("Custom agent reporting status without using fields")
 	a.TakeFood(15)
+	receivedMsg := a.Base.ReceiveMessage()
```
Comment (JaafarRammal) [inline pkg/agents/team1/agent2/agent.go:24]:
```suggestion
	receivedMsg := a.ReceiveMessage()
```

File: pkg/agents/team3/agentMain.go [hunk 7]
```diff
@@ -67,3 +65,13 @@ func (a *CustomAgent3) Run() {
...
+
+func (a *CustomAgent3) HandleAskHP(msg infra.AskMessage)                    {}
+func (a *CustomAgent3) HandleAskFoodTaken(msg infra.AskMessage)             {}
+func (a *CustomAgent3) HandleAskIntendedFoodTaken(msg infra.AskMessage)     {}
+func (a *CustomAgent3) HandleRequestLeaveFood(msg infra.RequestMessage)     {}
+func (a *CustomAgent3) HandleRequestTakeFood(msg infra.RequestMessage)      {}
+func (a *CustomAgent3) HandleResponse(msg infra.ResponseMessage)            {}
+func (a *CustomAgent3) HandleStateFoodTaken(msg infra.StateMessage)         {}
+func (a *CustomAgent3) HandleStateHP(msg infra.StateMessage)                {}
+func (a *CustomAgent3) HandleStateIntendedFoodTaken(msg infra.StateMessage) {}
```
Comment (JaafarRammal) [inline pkg/agents/team3/agentMain.go:77]:
```suggestion
```
These are already implemented in base agent so unless the team is using the handler there is no need to override with an empty function

File: pkg/agents/team6/config.go [hunk 8]
```diff
@@ -95,8 +94,18 @@ func (a *CustomAgent6) Run() {
...
+
+func (a *CustomAgent6) HandleAskHP(msg infra.AskMessage)                    {}
+func (a *CustomAgent6) HandleAskFoodTaken(msg infra.AskMessage)             {}
+func (a *CustomAgent6) HandleAskIntendedFoodTaken(msg infra.AskMessage)     {}
+func (a *CustomAgent6) HandleRequestLeaveFood(msg infra.RequestMessage)     {}
+func (a *CustomAgent6) HandleRequestTakeFood(msg infra.RequestMessage)      {}
+func (a *CustomAgent6) HandleResponse(msg infra.ResponseMessage)            {}
+func (a *CustomAgent6) HandleStateFoodTaken(msg infra.StateMessage)         {}
+func (a *CustomAgent6) HandleStateHP(msg infra.StateMessage)                {}
+func (a *CustomAgent6) HandleStateIntendedFoodTaken(msg infra.StateMessage) {}
```
Comment (JaafarRammal) [inline pkg/agents/team6/config.go:111]:
```suggestion
```
These are already implemented in base agent so unless the team is using the handler there is no need to override with an empty function

File: pkg/infra/baseMessage.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
+package infra
+
+//Define message types to enable basic protocols, voting systems ...etc
+
+type MessageType int
+
+// woonmoon(TODO): See if there's a more Go-friendly way to do this...
+// ideally Ask/State/Request would be three different Messagetypes
```
Comment (JaafarRammal) [inline pkg/infra/baseMessage.go:8]: I think this is currently good, I don't see a reason to have further inheritance for these three types (Ask / State / Request) unless there is a functional difference between them that justifies the inheritance

File: pkg/messages/askFoodTakenMessage.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	return msg
+}
+
+func (msg *AskFoodTakenMessage) Reply(senderFloor int, food int) infra.StateMessage {
+	reply := NewStateFoodTakenMessage(senderFloor, food)
+	return reply
+}
+
+func (msg *AskFoodTakenMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
```
Comment (JaafarRammal) [inline pkg/messages/askFoodTakenMessage.go:24]:
```suggestion
	return msg.MessageType()
```

File: pkg/messages/askFoodTakenMessage.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+func (msg *AskFoodTakenMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/askFoodTakenMessage.go:28]:
```suggestion
	return msg.SenderFloor()
```

File: pkg/messages/askHPMessage.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+	reply := NewStateHPMessage(senderFloor, hp)
+	return reply
+}
+
+func (msg *AskHPMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *AskHPMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/askHPMessage.go:28]:
```suggestion
	return msg.MessageType()
}

func (msg *AskHPMessage) SenderFloor() int {
	return msg.SenderFloor()
```

File: pkg/messages/askIntendedFoodIntakeMessage.go [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+	}
+	return msg
+}
+
+func (msg *AskIntendedFoodIntakeMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *AskIntendedFoodIntakeMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/askIntendedFoodIntakeMessage.go:23]:
```suggestion
	return msg.MessageType()
}

func (msg *AskIntendedFoodIntakeMessage) SenderFloor() int {
	return msg.SenderFloor()
```

File: pkg/messages/requestLeaveFoodMessage.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	reply := NewResponseMessage(senderFloor, response)
+	return reply
+}
+
+func (msg *RequestLeaveFoodMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *RequestLeaveFoodMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/requestLeaveFoodMessage.go:34]:
```suggestion
	return msg.MessageType()
}

func (msg *RequestLeaveFoodMessage) SenderFloor() int {
	return msg.SenderFloor()
```

File: pkg/messages/requestTakeFoodMessage.go [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+	reply := NewResponseMessage(senderFloor, response)
+	return reply
+}
+
+func (msg *RequestTakeFoodMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *RequestTakeFoodMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/requestTakeFoodMessage.go:34]:
```suggestion
	return msg.MessageType()
}

func (msg *RequestTakeFoodMessage) SenderFloor() int {
	return msg.SenderFloor()
```

File: pkg/messages/responseMessage.go [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+	}
+	return msg
+}
+
+func (msg *ResponseMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *ResponseMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/responseMessage.go:25]:
```suggestion
	return msg.MessageType()
}

func (msg *ResponseMessage) SenderFloor() int {
	return msg.SenderFloor()
```

File: pkg/messages/stateFoodTakenMessage.go [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+func (msg *StateFoodTakenMessage) Statement() int {
+	return msg.food
+}
+
+func (msg *StateFoodTakenMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *StateFoodTakenMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/stateFoodTakenMessage.go:29]:
```suggestion
	return msg.MessageType()
}

func (msg *StateFoodTakenMessage) SenderFloor() int {
	return msg.SenderFloor()
```

File: pkg/messages/stateHPMessage.go [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+	// TODO (woonmoon): GET RID OF THIS CONVERSION by supporting pseudo-union return types of Statement() function
+	return msg.hp
+}
+
+func (msg *StateHPMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *StateHPMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/stateHPMessage.go:33]:
```suggestion
	return msg.MessageType()
}

func (msg *StateHPMessage) SenderFloor() int {
	return msg.SenderFloor()
```

File: pkg/messages/stateIntendedFoodIntakeMessage.go [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+func (msg *StateIntendedFoodIntakeMessage) Statement() int {
+	return msg.intendedFood
+}
+
+func (msg *StateIntendedFoodIntakeMessage) MessageType() infra.MessageType {
+	return msg.baseMessage.MessageType()
+}
+
+func (msg *StateIntendedFoodIntakeMessage) SenderFloor() int {
+	return msg.baseMessage.SenderFloor()
```
Comment (JaafarRammal) [inline pkg/messages/stateIntendedFoodIntakeMessage.go:29]:
```suggestion
	return msg.MessageType()
}

func (msg *StateIntendedFoodIntakeMessage) SenderFloor() int {
	return msg.SenderFloor()
```

**Review by JaafarRammal — CHANGES_REQUESTED** [review]
A few minor changes, LGTM otherwise! New Visitor design is 🚀 

I am not sure if there is a way to avoid implementing `MessageType()` and `SenderFloor()` for every message since they are returning the function call from the base, and if we remove them we should be able to inherently call the base function directly when we try to call it from the implementing message?

**Review by SaraFFdez — DISMISSED** [review]
LGTM. I like this new implementation.

File: pkg/agents/team3/agentMain.go (hunk 7 above)
Comment (woonmoon) [reply pkg/agents/team3/agentMain.go]: Done. However, making `Base` into an embedded field on all custom agents means we can potentially run into the problem of some agent teams not implementing handlers at all which means they could receive messages and not realize an empty function was being called.
This is relatively easy to check for in a PR or we could throw an error on the handlers on `Base` so I don't feel very strongly, but something to consider.

File: pkg/simulation/simAgent.go [hunk 20]
```diff
@@ -28,7 +24,8 @@ func (sE *SimEnv) generateInitialAgents(t *infra.Tower) {
...
+	abs := []AgentNewFunc{agent1.New, agent2.New}
```
[the remark this answers is not in the document]
Comment (woonmoon) [reply pkg/simulation/simAgent.go]: Done but left the previous configuration in a comment for users to run if they wanted to see how message-passing between two agents worked.

File: pkg/messages/stateHPMessage.go [hunk 21]
```diff
@@ -0,0 +1,26 @@
+package messages
+
+// TODO (woonmoon): HP SHOULD BE AN INT!!!
+// do this by supporting union return types on Statement() in
+// AskMessage interface
```
Comment (hussain2603) [inline pkg/messages/stateHPMessage.go:5]:
```suggestion
```
Isn't HP an int already?

File: pkg/messages/stateIntendedFoodIntakeMessage.go [hunk 22]
```diff
@@ -0,0 +1,30 @@
+package messages
+
```
Comment (hussain2603) [inline pkg/messages/stateIntendedFoodIntakeMessage.go:2]: I think this file might be slightly wrong. see suggested changes

File: pkg/messages/stateIntendedFoodIntakeMessage.go [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+type StateIntendedFoodIntakeMessage struct {
+	baseMessage  *BaseMessage
```
Comment (hussain2603) [inline pkg/messages/stateIntendedFoodIntakeMessage.go:4]:
```suggestion
	*BaseMessage
```

File: pkg/messages/stateIntendedFoodIntakeMessage.go [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+	intendedFood int
+}
+
+func NewStateIntendedFoodIntakeMessage(SenderFloor int, intendedFood int) *StateIntendedFoodIntakeMessage {
+	msg := &StateIntendedFoodIntakeMessage{
+		baseMessage:  NewBaseMessage(SenderFloor, StateIntendedFoodIntake),
+		intendedFood: intendedFood,
```
Comment (hussain2603) [inline pkg/messages/stateIntendedFoodIntakeMessage.go:11]:
```suggestion
		NewBaseMessage(SenderFloor, StateIntendedFoodIntake),
		intendedFood,
```

**Review by hussain2603 — DISMISSED** [review]
a few minor changes suggested. Otherwise LGTM, thank so much on this Gowoon. It looks really nice.

**Review by JaafarRammal — APPROVED** [review]

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

**Review by jzzheng22 — APPROVED** [review]
LGTM, but can you fix the git history so it doesn't include commits which aren't actually in this PR?

**Pull request merged**